### PR TITLE
[WIP] Syntax for constraint domains

### PR DESCRIPTION
**Pull request by joehuchette**
For example:
```jl
@polyconstraint(m, f ≥ γ, domain = x1 >= 1 && x2-x1^2 >= 0)
```

### CODE REVIEW

**Comment by blegat** [comment]
I love the syntax ! It is definitely better than `[x1 >= 1, x2-x1^2 >= 0]`.
I see your implementation even supports nested syntax like `x >= 0 && (x >= 1 && x2 >= 0)`.
For the internal representation of the basic semialgebraic set I think we could add the following to `MultivariatePolynomials`:
```
type AlgebraicSet{T} <: AbstractBasicSemialgebraicSet{T}
    p::Vector{VecPolynomial{T}}
end
```
which represents `x == 0 && y^2 == 0` by `AlgebraicSet{Int}([x, y^2])` and
```
type BasicSemialgebraicSet{T} <: AbstractBasicSemialgebraicSet{T}
    V::AlgebraicSet{T}
    p::Vector{VecPolynomial{T}}
end
```
which represents `x == 0 && x >= 1 && y-x^2 >= 0 && y^2 == 0` as
```
BasicSemialgebraicSet{Int}(
    AlgebraicSet{Int}([x, y^2]),
    [x-1, y-x^2]
)
```
The `BasicSemialgebraicSet` would support `addequality!` and `addinequality!` that `PolyJuMP` could use.
The user could either do `domain = x1 >= 1 && x2-x1^2 >= 0` or `domain=d::AbstractBasicSemialgebraicSet`.
When the user give `domain=Hypercube(x::Vector{PolyVar})` then `SumOfSquares.jl` can use a specialized method that is super fast (see https://github.com/JuliaOpt/SumOfSquares.jl/issues/12).

As you can see I have started my Julia Packages with 2 space indentation but I have noticed that every body uses 4 spaces indentation so I am currently in the process of changing it in every julia code I wrote. If you see 2 space indentation, don't hesitate to make a reindent commit ;)

**Comment by blegat** [comment]
I have added the support for AlgebraicSet and BasicSemialgebraicSet in MultivariatePolynomials: https://github.com/blegat/MultivariatePolynomials.jl/commit/d8cf94f7f7081680e36236f9e41f3d6ceae9b86b
I choose to enforce any equality/inequality to be `VecPolynomial`s of the same `eltype`. Do you think it is too restrictive ?

**Comment by blegat** [comment]
The more I think about it, the more I think that it is indeed too restrictive. Instead of having a field `Vector{VecPolynomial{T}}`, it might be better to have a field `Vector` which is initialized as `Vector{Any}` by default but the user can use a more specific, e.g. `Vector{VecPolynomial{Int}}` if he wishes

**Comment by joehuchette** [comment]
Nice, the algebraic set types make this cleaner. 

I'll make a separate PR with white space changes if you don't mind, that should hopefully be less messy.

File: src/macros.jl [hunk 1]
```diff
@@ -125,7 +125,29 @@ macro polyvariable(args...)
...
+        elseif sense == :(<=)
+            push!(domain, :($(x.args[3]) - $(x.args[2])))
+        elseif sense == :(==) # for equality constraint, do x >= 0 and x <= 0
+            push!(domain, :($(x.args[2]) - $(x.args[3])))
+            push!(domain, :($(x.args[3]) - $(x.args[2])))
+        else
+            error("in @polyconstraint: Unrecognized sense $(string(sense)) in domain specification")
+        end
+    elseif isexpr(x, :&&)
+        map!(t -> appendconstraints!(domain, t), x.args)
```
Comment (joehuchette) [inline src/macros.jl]: It shouldn't, but ``map`` should work as well.

**Pull request merged**